File: backend/app/services/github.py
from datetime import datetime, timezone

import httpx

from app.config import Settings
from app.models import Branch, Commit, Repository


MOCK_REPOSITORIES = [
    Repository(id=1, name="commit-to-blog", full_name="octo/commit-to-blog", default_branch="main"),
    Repository(id=2, name="portfolio-api", full_name="octo/portfolio-api", default_branch="develop"),
]

MOCK_BRANCHES = {
    "octo/commit-to-blog": [Branch(name="main"), Branch(name="feature/blog-draft")],
    "octo/portfolio-api": [Branch(name="develop"), Branch(name="feature/github-sync")],
}

MOCK_COMMITS = {
    "octo/commit-to-blog": [
        Commit(
            sha="a1b2c3d",
            message="Add GitHub repository selector",
            author="dev",
            committed_at=datetime(2026, 5, 10, 9, 0, tzinfo=timezone.utc),
            url="https://github.com/octo/commit-to-blog/commit/a1b2c3d",
        ),
        Commit(
            sha="d4e5f6a",
            message="Generate blog draft from selected commits",
            author="dev",
            committed_at=datetime(2026, 5, 11, 13, 30, tzinfo=timezone.utc),
            url="https://github.com/octo/commit-to-blog/commit/d4e5f6a",
        ),
    ],
    "octo/portfolio-api": [
        Commit(
            sha="9ab8c7d",
            message="Create portfolio sync endpoint",
            author="dev",
            committed_at=datetime(2026, 5, 12, 8, 15, tzinfo=timezone.utc),
            url="https://github.com/octo/portfolio-api/commit/9ab8c7d",
        )
    ],
}


class GitHubService:
    def __init__(self, settings: Settings) -> None:
        self.settings = settings
        self.headers = {
            "Accept": "application/vnd.github+json",
            "Authorization": f"Bearer {settings.github_token}",
            "X-GitHub-Api-Version": "2022-11-28",
        }

    async def list_repositories(self) -> list[Repository]:
        if self.settings.use_mocks:
            return MOCK_REPOSITORIES

        async with httpx.AsyncClient(base_url="https://api.github.com", headers=self.headers, timeout=15) as client:
            response = await client.get("/user/repos", params={"sort": "updated", "per_page": 50})
            response.raise_for_status()
            return [
                Repository(
                    id=repo["id"],
                    name=repo["name"],
                    full_name=repo["full_name"],
                    default_branch=repo["default_branch"],
                )
                for repo in response.json()
            ]

    async def list_branches(self, repository_full_name: str) -> list[Branch]:
        if self.settings.use_mocks:
            return MOCK_BRANCHES.get(repository_full_name, [Branch(name="main")])

        async with httpx.AsyncClient(base_url="https://api.github.com", headers=self.headers, timeout=15) as client:
            response = await client.get(f"/repos/{repository_full_name}/branches")
            response.raise_for_status()
            return [Branch(name=branch["name"]) for branch in response.json()]

    async def list_commits(self, repository_full_name: str, branch: str) -> list[Commit]:
        if self.settings.use_mocks:
            return MOCK_COMMITS.get(repository_full_name, [])

        async with httpx.AsyncClient(base_url="https://api.github.com", headers=self.headers, timeout=15) as client:
            response = await client.get(f"/repos/{repository_full_name}/commits", params={"sha": branch, "per_page": 20})
            response.raise_for_status()
            commits = []
            for item in response.json():
                commit = item["commit"]
                commits.append(
                    Commit(
                        sha=item["sha"],
                        message=commit["message"],
                        author=commit["author"]["name"],
                        committed_at=datetime.fromisoformat(commit["author"]["date"].replace("Z", "+00:00")),
                        url=item["html_url"],
                    )
                )
            return commits

    async def selected_commits(self, repository_full_name: str, branch: str, commit_shas: list[str]) -> list[Commit]:
        commits = await self.list_commits(repository_full_name, branch)
        selected = [commit for commit in commits if commit.sha in commit_shas]
        return selected or commits[:1]

File: server/src/services/github.js
const GITHUB_API = 'https://api.github.com'

function getToken() {
  const token = process.env.GITHUB_TOKEN
  if (!token) {
    const err = new Error('GITHUB_TOKEN is not set. Add it to server/.env')
    err.code = 'MISSING_TOKEN'
    err.status = 500
    throw err
  }
  return token
}

async function githubFetch(path, { params } = {}) {
  const url = new URL(GITHUB_API + path)
  if (params) {
    for (const [k, v] of Object.entries(params)) {
      if (v !== undefined && v !== null) url.searchParams.set(k, v)
    }
  }

  const res = await fetch(url, {
    headers: {
      Authorization: `Bearer ${getToken()}`,
      Accept: 'application/vnd.github+json',
      'X-GitHub-Api-Version': '2022-11-28',
    },
  })

  if (!res.ok) {
    const body = await res.text().catch(() => '')
    const err = new Error(`GitHub ${res.status}: ${body.slice(0, 200)}`)
    err.status = res.status
    if (res.status === 401) err.code = 'INVALID_TOKEN'
    else if (res.status === 403) err.code = 'RATE_LIMIT_OR_FORBIDDEN'
    else if (res.status === 404) err.code = 'NOT_FOUND'
    throw err
  }

  return res.json()
}

function normalizeRepo(r) {
  return {
    id: r.id,
    owner: r.owner.login,
    name: r.name,
    fullName: r.full_name,
    defaultBranch: r.default_branch,
    description: r.description,
    updatedAt: r.updated_at,
    private: r.private,
  }
}

function normalizeBranch(b) {
  return {
    name: b.name,
    sha: b.commit.sha,
  }
}

function normalizeCommitSummary(c) {
  return {
    sha: c.sha,
    message: c.commit.message,
    author: {
      name: c.commit.author?.name ?? 'unknown',
      date: c.commit.author?.date ?? null,
    },
    url: c.html_url,
  }
}

function normalizeCommitDetail(c) {
  return {
    sha: c.sha,
    message: c.commit.message,
    author: { name: c.commit.author?.name ?? 'unknown' },
    date: c.commit.author?.date ?? null,
    files: (c.files ?? []).map((f) => ({
      filename: f.filename,
      status: f.status,
      additions: f.additions,
      deletions: f.deletions,
      patch: f.patch,
    })),
  }
}

export async function listRepos() {
  const data = await githubFetch('/user/repos', {
    params: { per_page: 100, sort: 'updated' },
  })
  return data.map(normalizeRepo)
}

export async function listBranches(owner, repo) {
  const data = await githubFetch(
    `/repos/${encodeURIComponent(owner)}/${encodeURIComponent(repo)}/branches`,
    { params: { per_page: 100 } },
  )
  return data.map(normalizeBranch)
}

export async function listCommits(owner, repo, branch) {
  const data = await githubFetch(
    `/repos/${encodeURIComponent(owner)}/${encodeURIComponent(repo)}/commits`,
    { params: { sha: branch, per_page: 20 } },
  )
  return data.map(normalizeCommitSummary)
}

export async function getCommit(owner, repo, sha) {
  const data = await githubFetch(
    `/repos/${encodeURIComponent(owner)}/${encodeURIComponent(repo)}/commits/${sha}`,
  )
  return normalizeCommitDetail(data)
}
